server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -47,12 +71,20 @@ public StoredFieldsSpec storedFieldsSpec() {
 
             @Override
             public void process(HitContext hitContext) throws IOException {
-                Map<String, DocumentField> documentFields = fieldFetcher.fetch(hitContext.source(), hitContext.docId());
-                SearchHit hit = hitContext.hit();
-                for (Map.Entry<String, DocumentField> entry : documentFields.entrySet()) {
-                    hit.setDocumentField(entry.getKey(), entry.getValue());
-                }
+                final FieldFetcher.DocAndMetaFields fields = fieldFetcher.fetch(hitContext.source(), hitContext.docId());
```

Comment:
In order to simplify things, I would propose that we make no changes to `FieldFetcher`, and instead create two instances of it, one for ordinary fields (like we already do) and a second one for metadata fields. The choice of where to put the fields that each fetcher outputs can be made in `FetchFieldsPhase`. This simplifies things in that it removes the need to distinguish between metadata and not in FieldFetcher and removes the need for the `fetchStoredFields` flag too.